### PR TITLE
./configure.py --terminate-on-asserts

### PR DESCRIPTION
**Pull request by reneme**
As [discussed](https://github.com/randombit/botan/discussions/2978), this adds `./configure.py --terminate-on-asserts` causing the library to `std::abort` when an assertion is not met.

The code base seems to use `BOTAN_ASSERT*` in some places where `BOTAN_ARG_CHECK` or `BOTAN_STATE_CHECK` would be more appropriate. Hence, I enabled the new option for the coverage CI run in the hope to catch most of them.

### CODE REVIEW

File: src/lib/utils/assert.cpp
Comment (randombit) [inline src/lib/utils/assert.cpp:11]: Is the `cstdlib` required in the general case or only with `BOTAN_TERMINATE_ON_ASSERTS`?

**Review by randombit — APPROVED** [review]
LGTM. Still a few CI failures which I assume are missing cases but once CI is green good to merge

**Comment by reneme** [comment]
The assertion in the coverage build job seems like an actual bug, actually.

There are a few BoGo tests that send a prefix string to the TLS server under test. E.g. BoGo test `HttpGET` sends `GET / HTTP/1.0\n"` and expects the TLS server under test to notice that and fail. The TLS 1.2 record implementation detects this situation and throws a `TLS_Exception`:

https://github.com/randombit/botan/blob/15228710faef1b85aaad0cad4f207b8b6dfcdf74/src/lib/tls/tls12/tls_record.cpp#L361-L368

The TLS channel then catches this exception and proceeds to sending a TLS alert in response. `TLS_Channel::send_alert()` used to swallow all exceptions that happened while trying to send an alert:

https://github.com/randombit/botan/blob/15228710faef1b85aaad0cad4f207b8b6dfcdf74/src/lib/tls/tls12/tls_channel_impl_12.cpp#L573-L577

But now, the `sequence_numbers()` accessor is called downstream, causing an assertion because the server didn't actually finish initializing and `m_sequence_numbers` is still a nullptr:

https://github.com/randombit/botan/blob/15228710faef1b85aaad0cad4f207b8b6dfcdf74/src/lib/tls/tls12/tls_channel_impl_12.cpp#L548-L552

I see two possible fixes:

* initialize `m_sequence_numbers` before any data is processed, or
* detect this situation and handle it

But is it actually legal for a TLS server to send a TLS alert as their first flight?

**Comment by reneme** [comment]
> But is it actually legal for a TLS server to send a TLS alert as their first flight?

At least for Botan's (D)TLS 1.2 implementation it doesn't seem to be legal for the TLS server to send messages before it successfully received a Client Hello. Only after that do we know whether we speak TLS or DTLS.

Hence, let's just not send TLS alerts in the very early error case described above: https://github.com/randombit/botan/pull/2980/commits/9c70d9d268241120104ad780bb3783e909a6750a

File: src/lib/pubkey/dsa/dsa.cpp
Comment (reneme) [inline src/lib/pubkey/dsa/dsa.cpp:41]: That is suppose to fix asserts in `PK_Verifier` for a DSA public key in the `cert` corpus/fuzzer. I left the assert untouched and instead introduce sanity checks at construction time of the DSA public key.
@randombit I could use a little help with this, as I'm not sure whether other things should also be validated here.

By the way: The currently still failing assert in the `pkcs8` fuzzer is related to a DSA private key:

```
* thread #1, queue = 'com.apple.main-thread', stop reason = signal SIGABRT
  * frame #0: 0x00007ff81db8e00e libsystem_kernel.dylib`__pthread_kill + 10
    frame #1: 0x00007ff81dbc41ff libsystem_pthread.dylib`pthread_kill + 263
    frame #2: 0x00007ff81db0fd24 libsystem_c.dylib`abort + 123
    frame #3: 0x00000001009a9778 pkcs8`Botan::assertion_failure(char const*, char const*, char const*, char const*, int) + 696
    frame #4: 0x000000010048ed39 pkcs8`Botan::Montgomery_Exponentation_State::exponentiation(Botan::BigInt const&, unsigned long) const + 1001
    frame #5: 0x000000010048fc46 pkcs8`Botan::monty_execute(Botan::Montgomery_Exponentation_State const&, Botan::BigInt const&, unsigned long) + 22
    frame #6: 0x0000000100651218 pkcs8`Botan::DL_Group_Data::power_g_p(Botan::BigInt const&, unsigned long) const + 72
    frame #7: 0x000000010065127b pkcs8`Botan::DL_Group::power_g_p(Botan::BigInt const&, unsigned long) const + 59
    frame #8: 0x0000000100660e4f pkcs8`Botan::DSA_PrivateKey::DSA_PrivateKey(Botan::AlgorithmIdentifier const&, std::__1::vector<unsigned char, Botan::secure_allocator<unsigned char> > const&) + 767
    frame #9: 0x000000010080913c pkcs8`std::__1::__unique_if<Botan::DSA_PrivateKey>::__unique_single std::__1::make_unique<Botan::DSA_PrivateKey, Botan::AlgorithmIdentifier const&, std::__1::vector<unsigned char, Botan::secure_allocator<unsigned char> > const&>(Botan::AlgorithmIdentifier const&, std::__1::vector<unsigned char, Botan::secure_allocator<unsigned char> > const&) + 76
    frame #10: 0x000000010080851b pkcs8`Botan::load_private_key(Botan::AlgorithmIdentifier const&, std::__1::vector<unsigned char, Botan::secure_allocator<unsigned char> > const&) + 1003
    frame #11: 0x000000010083fe17 pkcs8`Botan::PKCS8::(anonymous namespace)::load_key(Botan::DataSource&, std::__1::function<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > ()> const&, bool) + 471
    frame #12: 0x0000000100840587 pkcs8`Botan::PKCS8::load_key(Botan::DataSource&) + 231
    frame #13: 0x0000000100002c8f pkcs8`fuzz(unsigned char const*, unsigned long) + 255
    frame #14: 0x0000000100002b8a pkcs8`LLVMFuzzerTestOneInput + 26
    frame #15: 0x0000000100003483 pkcs8`(anonymous namespace)::fuzz_files(char**) + 899
    frame #16: 0x0000000100002eb7 pkcs8`main + 231
    frame #17: 0x0000000101df551e dyld`start + 462
```

File: src/lib/pubkey/dsa/dsa.cpp
Comment (reneme) [inline src/lib/pubkey/dsa/dsa.cpp:67]: I have a hunch that this check should better be constant time, because:

https://github.com/randombit/botan/blob/15228710faef1b85aaad0cad4f207b8b6dfcdf74/src/lib/math/numbertheory/monty_exp.cpp#L102-L105

File: src/lib/pubkey/dsa/dsa.cpp
Comment (randombit) [reply src/lib/pubkey/dsa/dsa.cpp:41]: I think this assert would trigger if you load a DSA key with a secret key that is greater than the group order.

This likely affects other key types as well.

I think we need to rethink key validation somewhat. Right now all validation is dumped on `check_key`, because some of the checks (testing primality and ECC point validity) are expensive. But most are simple range checks and could be cheaply done during construction or deserialization.

This is resolved with 1735f95 right?

File: src/lib/pubkey/dsa/dsa.cpp
Comment (randombit) [reply src/lib/pubkey/dsa/dsa.cpp:67]: I think this is safe enough to not be constant time in this context. `bits` itself is constant time, and if we reject that means the key itself is just manifest invalid. So there is no real info leak.

We should check `x <= q` though rather than just the bit size, since `x == q` would be accepted, but isn't a valid DSA key.

**Review by randombit — APPROVED** [review]
LGTM, modulo changing DSA key check to use value instead of bit length.

File: src/lib/pubkey/dsa/dsa.cpp
Comment (reneme) [reply src/lib/pubkey/dsa/dsa.cpp:69]: Nevertheless, I propose this is something for a follow-up PR. I wanted to keep this to the minimum to just start enforcing assertions in the first place.

**Comment by reneme** [comment]
I rebased to master to let the CI run once again after this was stale for more than a month. Once that comes back clean, I'll merge.

**Pull request merged**
